### DISCUSSION TITLE
Creating a preset from another

### DISCUSSION THREAD

**Discussion by faassen** [alvis/presetter#5] [Ideas]
Here's another scenario. We have a situation where we have plain packages, we have shared React component libraries, and we have full-fledged applications. Each of these share a lot of behavior: eslint (which we have in a reusable library anyway), tsconfig, a bunch of scripts etc, can be largely the same. Each layer just wants to add more, or possibly override a few files or settings in files.

How could presetter help to inherit/compose a preset from others so that we have such reuse?

**Comment by faassen** [comment]
Even though inheritance tends to be something I avoid, some kind of inheritance style pattern does seem to be appropriate in this case. If I base my preset off another preset, I expect:

* any config file or script I don't override to be there

* if I supply variables they should override the variables supplied by the previous preset

* if I supply a config file I either want to override it entirely or merge mine into the other, overriding any lines I have

**Reply by alvis** [reply]
Not being able to pick and match presets was definitely a pain in v2!
I personally have it enough and made a big change in [v3](https://github.com/alvis/presetter/tree/v3).

In v3, you can specify a list of presets in `.presetterrc.json`.
For example, if you want to combine the configurations provided by `presetter-preset` and `presetter-preset-react`,
you can specify them in `.presetterrc.json` as
```json
{
  "preset": ["presetter-preset", "presetter-preset-react"],
  "config": {
     "e.g. eslint": {
       "rules": {
         "rule-to-be-overridden": "your final say"
       }
     }
  }
}
```

All the configurations will be merged (or overridden in case of merge conflict) in the order of the preset list. i.e. some config from "presetter-preset-react" can override those from "presetter-preset" if there's some conflicts. Further, you always have the ultimate say via `config` in `.presetterrc.json` if you want to override the composed configurations as usual.

I first made the change via inheritance (e.g. `presetter-preset-react` includes `presetter-preset-web`). It's the easiest way to implement the feature, but the problem soon I faced is that all the dependencies must be linear and can't mix and match.

There are some documentation needed to be done, but I should be able to release a beta by tomorrow. Hope you can try and give some feedbacks. ;)
